src/envoy/http/path_matcher/filter.cc

```diff
@@ -69,27 +70,29 @@ Envoy::Http::FilterHeadersStatus Filter::decodeHeaders(
 
   std::string method(headers.Method()->value().getStringView());
   std::string path(headers.Path()->value().getStringView());
-  const std::string* operation = config_->findOperation(method, path);
-  if (operation == nullptr) {
+  const PathMatcherRule* rule = config_->findRule(method, path);
+  if (rule == nullptr) {
     rejectRequest(Envoy::Http::Code(404),
                   "Path does not match any requirement URI template.");
     return Envoy::Http::FilterHeadersStatus::StopIteration;
   }
 
-  ENVOY_LOG(debug, "matched operation: {}", *operation);
+  const absl::string_view operation = rule->operation();
+  ENVOY_LOG(debug, "matched operation: {}", operation);
   Envoy::StreamInfo::FilterState& filter_state =
       *decoder_callbacks_->streamInfo().filterState();
-  utils::setStringFilterState(filter_state, utils::kOperation, *operation);
+  utils::setStringFilterState(filter_state, utils::kOperation, operation);
+
+  if (rule->has_path_parameter_extraction()) {
+    const PathParameterExtractionRule& param_rule =
+        rule->path_parameter_extraction();
 
-  const PathParameterExtractionRule* path_param_extraction =
-      config_->needParameterExtraction(*operation);
-  if (path_param_extraction != nullptr) {
     std::vector<VariableBinding> variable_bindings;
-    config_->findOperation(method, path, &variable_bindings);
+    config_->findRule(method, path, &variable_bindings);
```

Comment:
Unrelated, but I noticed we duplicate the loop in the path matcher tree to extract variable bindings.

@qiwzhang is there any reason for the duplication? Can we just always extract them on line 73, even if we won't use them? 

Code would be cleaner. Performance may be a little slower for the majority of requests, but I doubt it will affect much.

Reply:
There are two different "find" methods, one without variable binding and one with.  the former is faster, without pass in so many parameters, without pass in segments.  Most methods not need to extract parameters, we want them to use faster call.
for the one that require parameter extraction, it is ok to call "find" twice.

Reply:
Ack